src/filters/mod.rs

```diff
@@ -65,45 +67,189 @@ fn check_args_len(
+//     offset: Value,
+//     length: Option<Value>,
+// }
+
+#[derive(Debug)]
+pub struct SliceFilter {
+    args: SliceParameters,
+}
+impl Filter for SliceFilter {
+    fn filter(&self, input: &Value, context: &Context) -> Result<Value> {
+        let args = self.args.evaluate(context)?;
+        
+        let offset = args.offset
+            .as_scalar()
+            .and_then(Scalar::to_integer)
+            .ok_or_else(|| invalid_argument(0, "Whole number expected"))?;
+        let offset = offset as isize;
+
+        let length = args.length
+            .unwrap_or(Value::scalar(1))
+            .as_scalar()
+            .and_then(Scalar::to_integer)
+            .ok_or_else(|| invalid_argument(0, "Whole number expected"))?;
+        if length < 1 {
+            return Err(invalid_argument(1, "Positive number expected"));
+        }
+        let length = length as isize;
+
+        if let Value::Array(ref input) = *input {
+            let (offset, length) = canonicalize_slice(offset, length, input.len());
+            Ok(Value::array(
+                input
+                    .iter()
+                    .skip(offset as usize)
+                    .take(length as usize)
+                    .cloned(),
+            ))
+        } else {
+            let input = input.to_str();
+            let (offset, length) = canonicalize_slice(offset, length, input.len());
+            Ok(Value::scalar(
+                input
+                    .chars()
+                    .skip(offset as usize)
+                    .take(length as usize)
+                    .collect::<String>(),
+            ))
+        }
     }
 }
 
-pub fn downcase(input: &Value, args: &[Value]) -> FilterResult {
-    check_args_len(args, 0, 0)?;
+#[derive(Clone)]
 
-    let s = input.to_str();
-    Ok(Value::scalar(s.to_lowercase()))
+pub struct SliceFilterParser;
+impl ParseFilter for SliceFilterParser {
+    fn parse(&self, args: FilterArguments) -> Result<Box<Filter>> {
+        let args = SliceParameters::new(args)?;
+        Ok(Box::new(SliceFilter { args }))
+    }
+}
+// TODO generate FilterReflection implementation from SliceParameters atributes
+impl FilterReflection for SliceFilterParser {
+    fn name(&self) -> &'static str {
+        "slice"
+    }
+    fn description(&self) -> &'static str {
+        "Takes a slice of a given string or array."
+    }
+    // Not sure on return type
+    // Goal is to return name and description
+    fn required_parameters(&self) -> &'static [(&'static str, &'static str)] {
+        &[("offset", "The offset of the slice.")]
+    }
+    fn optional_parameters(&self) -> &'static [(&'static str, &'static str)] {
+        &[("length", "The length of the slice.")]
+    }
+    // TODO does liquid have positional-only or keyword-only parameters?
+    fn positional_parameters(&self) -> &'static [(&'static str, &'static str)] {
+        &[("offset", "The offset of the slice."), ("length", "The length of the slice.")]
```

Comment:
The idea for `positional_parameters` is "positional-only", like pythons `*args`, so we shouldn't need to duplicate them here.

Reply:
Related:

> In my opinion, the keyword arguments should just be skipped in the positional count (so in filter: 1, a:2, 3, 3 would be the 2nd positional arg).
>
> Another thing that I've been thinking is: is it really needed to have both positional and keyword args in the same filter? Is there any liquid/jekyll/other filter that does that? Maybe we could just allow either positional or keyword arguments inside a filter, exclusively (FilterArguments would then be an enum instead).

The problem is I can't find any documentation on the semantics.  The only things I can find
- https://help.shopify.com/en/themes/liquid/filters/additional-filters#default_pagination
- https://github.com/Shopify/liquid/blob/master/test/integration/filter_test.rb#L140

I'm unsure what Liquid's semantics are.  My guess is they are Ruby's semantics.

I can speak for python's semantics

```python
def function(arg1, arg2, *pos, kwd1, **kwds):
    pass
```
- `arg1` and `arg2` can be referenced by position or name
- `kwd1` can only be referenced by name
- `pos` gobbles up all positional arguments after `arg2`
- `kwds` gobbles up any named arguments besides `arg1`, `arg2`, and `kwd1`.
- Once keyword arguments start, no more positional arguments can be specified

It was with the assumption of Python's semantics that I wrote the proposal (with TODOs about verifying them)

So looking at the Ruby to try to guess whats going on
- Parsing logic: https://github.com/Shopify/liquid/blob/master/lib/liquid/variable.rb#L96
  - Looks like keyword arguments are intermixed with positional arguments
- Preparing arguments for function: https://github.com/Shopify/liquid/blob/master/lib/liquid/variable.rb#L118
  - Looks like kwargs are made the last parameter, if I'm understanding the Ruby
- The actual call
  - https://github.com/Shopify/liquid/blob/master/lib/liquid/variable.rb#L84
  - https://github.com/Shopify/liquid/blob/master/lib/liquid/context.rb#L85
  - https://github.com/Shopify/liquid/blob/master/lib/liquid/strainer.rb#L54
  - So in the end, this depends on Ruby's semantics.

Reply:
So I've been doing some tests with Ruby on [Repl.it](https://repl.it/@Pedro_GoncaloGo/Ruby-positional-and-keyword-arguments-test) and I got to some interesting conclusions:
- An argument is either Positional-only or Keyword-only. At least, I wasn't able to find any syntax that would allow a parameter to be given either as a positional or keyword.
  - This means `FilterReflection` could instead have `positional_parameters` and `keyword_parameters`, returning `(name, description, is_optional)`, where `is_optional` could be a bool or maybe an enum.
  - We could take this even further and simply make a `parameters` function that returns `(name, description, positional_or_keyword, is_optional)`.
- Just like python, no positional args are allowed once keyword args start. As you said, though, ruby liquid filters bypass that, by automatically putting all kw-args in the end of the function call.
  - I think we should separate them. If we had something like `Vec<(Option<&str>, &str)>`, it would be harder to iterate on the positional arguments by their order (if they had any named parameter in the middle it would have to be skipped).
- In the case where the function definition only declares positional arguments and not even a single keyword argument, if we pass in keyword arguments, they will all be merged into a hashmap and work as if we passed in a single positional argument (that happens to be a hashmap) instead. This is the closest I could get to make an argument not kw-only or pos-only.
  - It seems to me that it's such a rare case that it might be unintended. I don't think ruby liquid would want us to pass keyword arguments to a filter that asked for an `Object` positional argument, for instance. So I think we should just ignore this edge case.

Given this, you can go to the link I gave above, if you want. It leads directly to a "project" with my test code and some comments about those conclusions I just pointed out. You can play with it and maybe find something I missed)

Reply:
From this, some thoughts
- It seems like our macros will need field attributes to declare a field as positional or keyword
- Should we allow "catch all" attributes that are `Vec` and `HashMap` in case a filter author wants to do something very custom?  Nevermind, I guess they could skip the macro in this case.

> ` returning (name, description, is_optional), where is_optional could be a bool or maybe an enum.`

Another option is `Option<&'static str>` where the `str` is a description of the default value.  Except will everything have a default?  Hmm.

> We could take this even further and simply make a parameters function that returns `(name, description, positional_or_keyword, is_optional)`.

Not sure what my thoughts are.  If we do this, `positional_or_keyword` should be an `enum` I think.

Reply:
> * It seems like our macros will need field attributes to declare a field as positional or keyword

Yes, this is actually where one of my doubts arises. Because the vast majority of filters will have positional parameters (to the point that keyword parameters are not even very well documented in liquid), I was thinking of making positional arguments the default (so, if no attribute was present, it would assume it's a positional argument. If keyword attr was present, it would be a keyword argument).

Do you like that or would you prefer more explicitness, forcing to always use an attribute?

> Not sure what my thoughts are. If we do this, `positional_or_keyword` should be an `enum` I think.

Yeah, that was exactly my intention. The issue I see with this possibility is that we would then have a 4-elements tuple `(&str, &str, ParameterType, bool)`, which could make it hard to understand the meaning of the return type.

> > ` returning (name, description, is_optional), where is_optional could be a bool or maybe an enum.`
> 
> Another option is `Option<&'static str>` where the `str` is a description of the default value. Except will everything have a default? Hmm.

I think this would aggravate the problem with the tuple structure. Having a `(&str, &str, Option<&str>)`  would be pretty confusing.

Reply:
> Yes, this is actually where one of my doubts arises. Because the vast majority of filters will have positional parameters (to the point that keyword parameters are not even very well documented in liquid), I was thinking of making positional arguments the default (so, if no attribute was present, it would assume it's a positional argument. If keyword attr was present, it would be a keyword argument).

Having positional be default makes sense.

> Yeah, that was exactly my intention. The issue I see with this possibility is that we would then have a 4-elements tuple (&str, &str, ParameterType, bool), which could make it hard to understand the meaning of the return type.

We could have a struct for this so that the fields are named.  In the python world, I sometimes refer to tuples as anonymous structs to help people understand their difference from lists.  We'd just be making the return type not-anonymous.

Reply:
Keeping positional arguments and keyword arguments in separated functions maybe would be more useful. Because the user of the function will probably need them separated (for example to have two different documentation sections, one for positional and another for keyword). By mixing them we are just "wasting" performance because the user would then have to do the sorting (of positional vs keyword) instead of it being already statically done.